From e7ba95893f424ca835def3a21b1e132ed845e7ad Mon Sep 17 00:00:00 2001
From: Md Sadre Alam <quic_mdalam@quicinc.com>
Date: Fri, 17 Nov 2023 15:33:57 +0530
Subject: [PATCH] crypto: qce - Fix potential memleak problem for aead

Fix poetential memleak problem for aead algorithm.
In qce_aead_ccm_prepare_buf_assoclen() API 'adata' is
getting alloacted dynamically using kmalloc(), this
pointer is only used in qce_aead_ccm_prepare_buf_assoclen()
API only so before returing from this API free the 'adata'.

qce_aead_ccm_prepare_buf_assoclen
           |
           |
	kfree(adata)

return 0;

If we will not free this memory and if CONFIG_DEBUG_KMEMLEAK=y
is enabled then we can see the below stack trace.

root@OpenWrt:/# cat /sys/kernel/debug/kmemleak
unreferenced object 0xffffff803e26d880 (size 128):
  comm "cryptomgr_test", pid 152, jiffies 4294937597 (age 3666.950s)
  hex dump (first 32 bytes):
    00 08 00 01 02 03 04 05 06 07 00 00 00 00 00 00  ................
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  backtrace:
    [<00000000b3183fa8>] __kmalloc+0x240/0x274
    [<000000007fc4da09>] qce_aead_async_req_handle+0x214/0x668
    [<000000007b1e856a>] qce_handle_queue+0x10c/0x15c
    [<00000000f1c6e9b9>] qce_async_request_enqueue+0x16c/0x17c
    [<00000000f5836e6b>] qce_aead_crypt+0x114/0x124
    [<00000000c70bf107>] qce_aead_encrypt+0x10/0x18
    [<00000000466b44ce>] crypto_aead_encrypt+0x20/0x30
    [<00000000e0e302d6>] test_aead_vec_cfg+0x2e4/0x71c
    [<0000000084e79bfa>] test_aead+0xd4/0x130
    [<0000000087315c2f>] alg_test_aead+0xd4/0x150
    [<0000000083afa2c4>] alg_test.part.13+0x28c/0x360
    [<000000002e047969>] alg_test+0x44/0x58
    [<00000000f9f93c14>] cryptomgr_test+0x24/0x4c
    [<00000000aad9fc24>] kthread+0x14c/0x15c
    [<00000000a358441b>] ret_from_fork+0x10/0x18

This is a snapshot of the commit
4be7f1a15772d044b5344cd916087d6e102e83d2
from win.linuxopenwrt.2.1 branch and
oss/kernel/linux-ipq-5.4 project.

Change-Id: I2de6a990a1b185a8c1290c041c4c16a9b247b9e9
Signed-off-by: Md Sadre Alam <quic_mdalam@quicinc.com>
---
 drivers/crypto/qce/aead.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/crypto/qce/aead.c b/drivers/crypto/qce/aead.c
index ca47859acab9..429da142c795 100644
--- a/drivers/crypto/qce/aead.c
+++ b/drivers/crypto/qce/aead.c
@@ -307,6 +307,8 @@ qce_aead_ccm_prepare_buf_assoclen(struct aead_request *req)
 		rctx->dst_sg = rctx->src_sg;
 	}
 
+	kfree(rctx->adata);
+
 	return 0;
 err_free:
 	sg_free_table(&rctx->src_tbl);
-- 
2.34.1

